test/testcne/cthread_test.c

```diff
@@ -81,10 +82,96 @@ print_data(int type)
     }
 }
 
+static int
+cthread_sema_tests(void)
+{
+    struct cthread_sema *s,
+        sema = {.blocked = 0, .sched = 0, .name = "sema_test", .orig_count = 0, .count = 0};
+    const char *sema_name    = "testcne semaphore";
+    struct timespec waittime = {.tv_sec = 5};
+    int ret                  = 0;
+
+    s = calloc(1, sizeof(struct cthread_sema));
+    if (!s) {
+        printf("Unable to allocate memory\n");
+        return -1;
+    }
+
+    s = &sema;
+
+    ret = cthread_sema_init(sema_name, &s, NULL);
+    if (ret < 0) {
+        ret = errno;
+        tst_error("Unable to initialize a semaphore, %s\n", strerror(errno));
+        /* return errno if this test fails due to permission error */
+        free(s);
+        return (ret == EPERM) || (ret == EACCES) ? ret : -1;
+    }
+
+    tst_ok("PASS --- TEST: Initialize semaphore\n");
+
+    ret = cthread_sema_reset(s);
+    if (ret < 0) {
+        tst_error("Unable to reset semaphore, %s\n", strerror(errno));
+        goto err;
+    }
+
+    tst_ok("PASS --- TEST: Reset semaphore\n");
+
+    ret = cthread_sema_timedwait(s, NULL, &waittime);
+    if (ret < 0) {
+        tst_error("Error waiting on semaphore, %s\n", strerror(errno));
+        goto err;
+    }
+
+    tst_ok("PASS --- TEST: Timed wait on semaphore\n");
+
+    ret = cthread_sema_signal(s);
+    if (ret < 0) {
+        tst_error("Error signaling semaphore, %s\n", strerror(errno));
+        goto err;
+    }
+
+    tst_ok("PASS --- TEST: Signal semaphore\n");
+
+    ret = cthread_sema_flush_no_sched(s);
+    if (ret < 0) {
+        tst_error("Error flushing semaphore w/o rescheduling threads, %s\n", strerror(errno));
+        goto err;
+    }
+
+    tst_ok("PASS --- TEST: Flush semaphore, w/o rescheduling threads\n");
+
+    ret = cthread_sema_flush(s);
+    if (ret < 0) {
+        tst_error("Error flushing semaphore, %s\n", strerror(errno));
+        goto err;
+    }
+
+    tst_ok("PASS --- TEST: Flush semaphore\n");
+
+    ret = cthread_sema_destroy(s);
+    if (ret < 0) {
+        tst_error("Unable to destroy semaphore, %s\n", strerror(errno));
+        goto err;
+    }
+
+    tst_ok("PASS --- TEST: Destroy semaphore\n");
+
+    free(s);
+    return 0;
+
+err:
+    free(s);
```

Comment:
Have a look at chread_sema_destroy() I believe it will accept NULL and you can call chread_sema_destroy(s) here before the free(s).